lib/Transforms/ValidateNoise/ValidateNoise.cpp

```diff
@@ -0,0 +1,176 @@
+#include "lib/Transforms/ValidateNoise/ValidateNoise.h"
+
+#include "lib/Analysis/DimensionAnalysis/DimensionAnalysis.h"
+#include "lib/Analysis/LevelAnalysis/LevelAnalysis.h"
+#include "lib/Analysis/NoiseAnalysis/BGV/NoiseByBoundCoeffModel.h"
+#include "lib/Analysis/NoiseAnalysis/NoiseAnalysis.h"
+#include "lib/Dialect/Secret/IR/SecretOps.h"
+#include "llvm/include/llvm/Support/Debug.h"  // from @llvm-project
+#include "mlir/include/mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"  // from @llvm-project
+#include "mlir/include/mlir/Analysis/DataFlow/DeadCodeAnalysis.h"  // from @llvm-project
+#include "mlir/include/mlir/Analysis/DataFlowFramework.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"     // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"           // from @llvm-project
+
+#define DEBUG_TYPE "ValidateNoise"
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_VALIDATENOISE
+#include "lib/Transforms/ValidateNoise/ValidateNoise.h.inc"
+
+struct ValidateNoise : impl::ValidateNoiseBase<ValidateNoise> {
+  using ValidateNoiseBase::ValidateNoiseBase;
+
+  // assume only one main func
+  // also assume max level at entry
+  // also assume first genericOp arg is secret
+  int getMaxLevel() {
+    int maxLevel = 0;
+    getOperation()->walk([&](func::FuncOp funcOp) {
+      funcOp->walk([&](secret::GenericOp genericOp) {
+        for (Value arg : genericOp.getBody()->getArguments()) {
+          maxLevel = getLevelFromMgmtAttr(arg);
+          break;
+        }
+      });
+    });
+    return maxLevel;
+  }
+
+  template <typename NoiseAnalysis>
+  LogicalResult validateNoiseForValue(
+      Value value, DataFlowSolver *solver,
+      const typename NoiseAnalysis::SchemeParamType &schemeParam) {
+    using NoiseModel = typename NoiseAnalysis::NoiseModel;
+    using NoiseLatticeType = typename NoiseAnalysis::LatticeType;
+    using LocalParamType = typename NoiseAnalysis::LocalParamType;
+
+    auto getLocalParam = [&](Value value) {
+      auto level = getLevelFromMgmtAttr(value);
+      auto dimension = getDimensionFromMgmtAttr(value);
+      return LocalParamType(&schemeParam, level, dimension);
+    };
+
+    auto secretness = isSecret(value, solver);
+    if (!secretness) {
+      return success();
+    }
+
+    const auto *noiseLattice = solver->lookupState<NoiseLatticeType>(value);
+    if (!noiseLattice || !noiseLattice->getValue().isInitialized()) {
+      return failure();
+    }
+
+    auto noiseState = noiseLattice->getValue();
+    auto localParam = getLocalParam(value);
+
+    int total = 0;
+    auto logqi = localParam.getSchemeParam()->getLogqi();
+    for (auto i = 0; i <= localParam.getCurrentLevel(); ++i) {
+      total += logqi[i];
+    }
+
+    auto bound = NoiseModel::toLogBound(localParam, noiseState);
+    auto boundString = NoiseModel::toLogBoundString(localParam, noiseState);
```

Comment:
```suggestion
    [[maybe_unused]] auto boundString = NoiseModel::toLogBoundString(localParam, noiseState);
```

You could also put this inside the debug below.